sklearn/linear_model/ridge.py

```diff
@@ -995,6 +993,100 @@ def _diag_dot(self, D, B):
             D = D[(slice(None), ) + (np.newaxis, ) * (len(B.shape) - 1)]
         return D * B
 
+    def _compute_gram(self, X, center=True):
+        """Computes centered Gram matrix.
+
+        Notes
+        -----
+        if center is True, compute
+        (X - X.mean(axis=0)).dot((X - X.mean(axis=0)).T)
+        else
+        X.dot(X.T)
+        """
+        if not center:
+            X_m = np.zeros(X.shape[1], dtype=X.dtype)
+            return safe_sparse_dot(X, X.T, dense_output=True), X_m
+        n = X.shape[0]
+        X_w = self._sqrt_sw_matrix.dot(X)
+        X_m, _ = mean_variance_axis(X_w, axis=0)
+        X_m = X_m * n / self._weight_sum
+        X_mX = self._sqrt_sw[:, None] * safe_sparse_dot(
+            X_m, X.T, dense_output=True)
+        X_mX_m = np.empty((X.shape[0], X.shape[0]), dtype=X.dtype)
+        X_mX_m[:, :] = np.dot(X_m, X_m)
+        X_mX_m = X_mX_m * self._sqrt_sw
+        X_mX_m = X_mX_m * self._sqrt_sw[:, None]
+        return (safe_sparse_dot(X, X.T, dense_output=True) + X_mX_m
+                - X_mX - X_mX.T, X_m)
+
+    def _compute_covariance(self, X, center=True):
+        """Computes centered Gram matrix.
+
+        Notes
+        -----
+        if center is True, compute
+        (X - X.mean(axis=0)).T.dot(X - X.mean(axis=0))
+        else
+        X.T.dot(X)
+        """
+        if not center:
+            X_m = np.zeros(X.shape[1], dtype=X.dtype)
+            return safe_sparse_dot(X.T, X, dense_output=True), X_m
+        n = X.shape[0]
+        X_w = self._sqrt_sw_matrix.dot(X)
+        X_m, _ = mean_variance_axis(X_w, axis=0)
+        X_m = X_m * n / self._weight_sum
+        return safe_sparse_dot(
+            X.T, X, dense_output=True) - self._weight_sum * np.outer(
+                X_m, X_m), X_m
```

Comment:
It might be a bit easier to parse visually

```python
        return (safe_sparse_dot(X.T, X, dense_output=True) - 
                self._weight_sum * np.outer(X_m, X_m),
                X_m)
```